File: platforms/windows/Dimmy.Windows/Services/FileLoadToMeetingService.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

namespace Dimmy.Windows.Services;

/// <summary>
/// Bridge between the Settings → "Transcribe a file" card and the
/// Meeting recap pipeline. Given a successful file-load transcript +
/// the source WAV path, this:
///   1. Mints a synthetic meeting directory under the same root the
///      live recorder uses (<c>%APPDATA%\dimmy\meetings\&lt;guid&gt;</c>),
///   2. Writes `transcripts.txt` in the same `[ELAPSED_MS ms] [LABEL] text`
///      grammar the recap LLM prompt expects (so the existing
///      <see cref="MeetingPostProcessService"/> works without a code
///      branch),
///   3. Writes `meta.json` with title/started_at/duration so the
///      Meeting → History list renders the new entry nicely,
///   4. Optionally copies the source WAV to `audio.wav` so playback
///      from the Done view works (small enough to be worth it for
///      most files; the user can pick a 95-min recording too and a
///      copy is fine — disk is cheap, we already keep history audio
///      under retention rules in the Rust core),
///   5. Hands off to <see cref="MeetingPostProcessService.RunRecapAsync"/>
///      which does the LLM call + writes `recap.md` / `actions.json`.
///
/// Single-writer rule is preserved: only this service writes to the
/// meeting dir on the file-load path; Rust still owns config.json and
/// the live-meeting writers.
/// </summary>
public static class FileLoadToMeetingService
{
    public sealed class Result
    {
        public bool Success { get; init; }
        public string Dir { get; init; } = "";
        public string? Error { get; init; }
    }

    /// <summary>
    /// Create a meeting dir for the file-load transcript and trigger
    /// the recap pipeline. Returns the dir + success/failure so the
    /// caller can deep-link the user into the Done view.
    /// </summary>
    public static async Task<Result> RunAsync(string sourceWavPath, string transcript)
    {
        if (string.IsNullOrWhiteSpace(transcript))
            return new Result { Success = false, Error = "empty transcript" };

        try
        {
            var meetingsRoot = Path.Combine(BuildInfo.ConfigDirPath, "meetings");
            Directory.CreateDirectory(meetingsRoot);

            var dirName = $"{DateTime.UtcNow:yyyyMMddTHHmmss}-{Guid.NewGuid():N}";
            var dir = Path.Combine(meetingsRoot, dirName);
            Directory.CreateDirectory(dir);

            // transcripts.txt — single line; the recap LLM prompt
            // tolerates a single `[0 ms] [<label>] <text>` line.
            var label = "file";
            var transcriptsPath = Path.Combine(dir, "transcripts.txt");
            await File.WriteAllTextAsync(
                transcriptsPath,
                $"[0 ms] [{label}] {transcript.Trim()}{Environment.NewLine}");

            // audio.wav — sidecar for the Done view (waveform + media
            // player). WAV sources can be copied verbatim; everything
            // else (m4a / mp3 / aac / flac / ogg) goes through the
            // Rust `dimmy_decode_audio_to_wav` FFI so the resulting
            // file is a real RIFF/WAVE container and the WAV-only
            // consumers downstream (WavPeaks.ReadPeaks,
            // MediaPlayerElement on some codec-poor systems) keep
            // working without per-format branches.
            double durationSecs = 0;
            try
            {
                if (File.Exists(sourceWavPath))
                {
                    var audioDest = Path.Combine(dir, "audio.wav");
                    var ext = Path.GetExtension(sourceWavPath);
                    bool isWav = !string.IsNullOrEmpty(ext)
                        && ext.Equals(".wav", StringComparison.OrdinalIgnoreCase);
                    if (isWav)
                    {
                        File.Copy(sourceWavPath, audioDest, overwrite: true);
                    }
                    else
                    {
                        int rc = await Task.Run(() =>
                            Dimmy.Windows.Interop.DimmyNative.dimmy_decode_audio_to_wav(
                                sourceWavPath, audioDest));
                        if (rc <= 0)
                        {
                            App.Log($"FileLoadToMeeting: decode-to-wav rc={rc} for {sourceWavPath}", "FileLoad");
                        }
                    }
                    durationSecs = Helpers.WavPeaks.ReadDurationSecs(audioDest);
                }
            }
            catch (Exception ex)
            {
                App.Log($"FileLoadToMeeting: audio copy/decode failed: {ex.Message}", "FileLoad");
            }

            // meta.json — the MeetingWindow history loader reads
            // `started_at` (string) + `duration_secs` (double) + `title`
            // (string). Title comes from the source filename so the user
            // can tell loaded files apart from live recordings at a
            // glance.
            var fname = string.IsNullOrEmpty(sourceWavPath)
                ? "Imported audio"
                : Path.GetFileNameWithoutExtension(sourceWavPath);
            var meta = new
            {
                title = $"File: {fname}",
                started_at = DateTime.Now.ToString("yyyy-MM-dd HH:mm"),
                duration_secs = durationSecs,
                origin = "file-load",
            };
            await File.WriteAllTextAsync(
                Path.Combine(dir, "meta.json"),
                JsonSerializer.Serialize(meta, new JsonSerializerOptions { WriteIndented = true }));

            App.Log($"FileLoadToMeeting: dir created at {dir}, transcript={transcript.Length}c, duration={durationSecs:F1}s", "FileLoad");
            Dimmy.Windows.Interop.DimmyNative.TrackEvent("meeting.imported_from_file");

            var recapResult = await MeetingPostProcessService.RunRecapAsync(dir, transcript);
            if (!recapResult.Success)
            {
                return new Result
                {
                    Success = false,
                    Dir = dir,
                    Error = recapResult.Error ?? "recap failed",
                };
            }

            App.Log($"FileLoadToMeeting: recap succeeded for dir={dir}", "FileLoad");
            return new Result { Success = true, Dir = dir };
        }
        catch (Exception ex)
        {
            App.Log($"FileLoadToMeeting: exc {ex.GetType().Name}: {ex.Message}", "FileLoad");
            return new Result { Success = false, Error = ex.Message };
        }
    }
}

File: platforms/macos/Dimmy/Services/FileLoadToMeetingService.swift
import Foundation

// MARK: - FileLoadToMeetingService
//
// Mac mirror of Win `Services/FileLoadToMeetingService.cs`. Bridges the
// File Load card's transcript to the Meeting recap pipeline.
//
// Given a successful local transcript + the source WAV path, this:
//   1. Mints a synthetic meeting directory under
//      `<configDir>/meetings/<timestamp>-<uuid>` — the same root the
//      live recorder writes to, so Meeting → History picks it up
//      automatically.
//   2. Writes `transcripts.txt` in the `[ELAPSED_MS ms] [LABEL] text`
//      grammar `MeetingPostProcessService.buildStructuredRecapPrompt`
//      expects — single line `[0 ms] [file] <text>` for a file-load.
//   3. Writes `meta.json` with `title`/`started_at`/`duration_secs` so
//      the History row renders the right subtitle + duration.
//   4. Best-effort copies the source WAV to `audio.wav` so the Done
//      view's audio scrubber works. Failure here doesn't abort the
//      recap — disk-full / permission errors just degrade to "recap
//      without playback".
//   5. Hands off to `MeetingPostProcessService.runRecap(dir:transcript:)`
//      which calls the LLM and persists `recap.md` + `actions.json`.
//
// Single-writer rule preserved: only this service writes to the
// meeting dir on the file-load path; Rust still owns config.json + the
// live-meeting writers.

enum FileLoadToMeetingService {

    struct Outcome {
        let dir: String
        let recapMarkdown: String?
        let error: String?
        var success: Bool { error == nil }
    }

    /// Run the full file-load → meeting pipeline. Blocking — call
    /// from a background thread (the LLM step inside
    /// `MeetingPostProcessService.runRecap` can take 10–60 s).
    ///
    /// `notionAutoSend` MUST be captured on the caller's main-actor
    /// thread before dispatch — Swift 6 strict concurrency rejects
    /// any read of `AppState.shared.notionAutoSend` (main-actor
    /// isolated) from a `DispatchQueue.global` closure as a hard
    /// error.
    static func run(sourceWavPath: String, transcript: String, notionAutoSend: Bool) -> Outcome {
        let trimmedTranscript = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedTranscript.isEmpty else {
            return Outcome(dir: "", recapMarkdown: nil, error: "empty transcript")
        }

        // Effective meetings dir (honours the user's meeting_storage_path
        // override) — never re-derive configDir/meetings here.
        guard let meetingsRoot = DimmyCore.shared.meetingsDirURL else {
            return Outcome(dir: "", recapMarkdown: nil, error: "meetings dir unavailable")
        }
        do {
            try FileManager.default.createDirectory(at: meetingsRoot, withIntermediateDirectories: true)
        } catch {
            return Outcome(dir: "", recapMarkdown: nil,
                           error: "could not create meetings root: \(error.localizedDescription)")
        }

        // Dir name shape mirrors the live worker's so the History
        // sort-by-modified-time keeps file-load entries inline with
        // recorded ones.
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withYear, .withMonth, .withDay, .withTime]
        let stamp = Date().formatted(.iso8601.year().month().day().time(includingFractionalSeconds: false))
            .replacingOccurrences(of: ":", with: "")
            .replacingOccurrences(of: "-", with: "")
        let dirName = "\(stamp)-\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
        let dir = meetingsRoot.appendingPathComponent(dirName, isDirectory: true)
        do {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        } catch {
            return Outcome(dir: dir.path, recapMarkdown: nil,
                           error: "could not create meeting dir: \(error.localizedDescription)")
        }

        // transcripts.txt — the recap prompt expects the
        // `[ELAPSED_MS ms] [LABEL] text` grammar (see
        // MeetingPostProcessService.buildStructuredRecapPrompt). A
        // file-load is one big line — the LLM still produces a useful
        // recap, just without intra-meeting time anchors.
        let transcriptsURL = dir.appendingPathComponent("transcripts.txt")
        let transcriptLine = "[0 ms] [file] \(trimmedTranscript)\n"
        do {
            try transcriptLine.write(to: transcriptsURL, atomically: true, encoding: .utf8)
        } catch {
            return Outcome(dir: dir.path, recapMarkdown: nil,
                           error: "could not write transcripts.txt: \(error.localizedDescription)")
        }

        // audio.wav copy + duration probe. Best-effort.
        var durationSecs: Double = 0
        if !sourceWavPath.isEmpty,
           FileManager.default.fileExists(atPath: sourceWavPath) {
            let dest = dir.appendingPathComponent("audio.wav")
            do {
                try? FileManager.default.removeItem(at: dest)
                try FileManager.default.copyItem(at: URL(fileURLWithPath: sourceWavPath), to: dest)
                durationSecs = WavPeaks.readDurationSecs(path: dest.path)
            } catch {
                NSLog("[FileLoadToMeeting] audio.wav copy failed: \(error.localizedDescription)")
            }
        }

        // meta.json — Meeting → History reads `started_at` (string)
        // + `duration_secs` (Double) + `title` (String). Title comes
        // from the source filename so the user can tell file-load
        // entries apart from live recordings at a glance.
        let baseName: String
        if sourceWavPath.isEmpty {
            baseName = "Imported audio"
        } else {
            baseName = URL(fileURLWithPath: sourceWavPath).deletingPathExtension().lastPathComponent
        }
        let metaURL = dir.appendingPathComponent("meta.json")
        let dateFmt = DateFormatter()
        dateFmt.dateFormat = "yyyy-MM-dd HH:mm"
        let meta: [String: Any] = [
            "title": "File: \(baseName)",
            "started_at": dateFmt.string(from: Date()),
            "duration_secs": durationSecs,
            "origin": "file-load",
        ]
        if let json = try? JSONSerialization.data(
            withJSONObject: meta, options: [.prettyPrinted]) {
            try? json.write(to: metaURL)
        }

        NSLog("[FileLoadToMeeting] dir=\(dir.path) transcript=\(trimmedTranscript.count)c duration=\(durationSecs)s")

        // Tag the synthetic meeting so the funnel can distinguish
        // "live recorded" from "imported audio". Win emits this
        // before kicking off the recap; we mirror so dashboards
        // are platform-agnostic.
        DimmyCore.shared.trackEvent("meeting.imported_from_file")

        // Hand off to the shared recap pipeline. `notionAutoSend`
        // was captured on the caller's main thread (see docstring).
        // We still gate it on `notionHasToken` here because the
        // token state is owned by DimmyCore and safe to read from
        // any thread (it's a thin FFI getter, not @MainActor).
        let effectiveAutoSend = notionAutoSend && DimmyCore.shared.notionHasToken
        let recapResult = MeetingPostProcessService.runRecap(
            dir: dir.path,
            transcript: trimmedTranscript,
            modelOverride: nil,
            notionAutoSend: effectiveAutoSend
        )
        switch recapResult {
        case .success(let res):
            NSLog("[FileLoadToMeeting] recap ok at \(dir.path)")
            return Outcome(dir: dir.path, recapMarkdown: res.recapMarkdown, error: nil)
        case .failure(let err):
            return Outcome(dir: dir.path, recapMarkdown: nil, error: "\(err)")
        }
    }
}
